src/pgduckdb_types.cpp

```diff
@@ -334,9 +334,29 @@ ConvertNumeric(const duckdb::Value &ddb_value, idx_t scale, NumericVar &result)
 	}
 }
 
+// A wrapper function to convert a string to a numeric value, used for `PostgresFunctionGuard` macro.
+static Datum
+ConvertStringToNumerics(Datum str, Datum typelem, Datum typmod) {
+	Datum pg_numeric = DirectFunctionCall3(numeric_in, str, typelem, typmod);
+	return pg_numeric;
+}
+
 static Datum
 ConvertNumericDatum(const duckdb::Value &value) {
 	auto value_type_id = value.type().id();
+
+	// Special handle duckdb VARINT type.
+	if (value.type().id() == duckdb::LogicalTypeId::VARINT) {
+		// The performant way to handle the translation is to parse VARINT out, here we leverage string conversion and
+		// parsing mainly for code simplicity.
+		const std::string value_str = value.ToString();
+		const Datum pg_numeric =
+		    PostgresFunctionGuard(ConvertStringToNumerics, /*str=*/CStringGetDatum(value_str.c_str()),
+		                          /*typelen=*/ObjectIdGetDatum(InvalidOid), /*typmod=*/Int32GetDatum(-1));
```

Comment:
I think it makes most sense to create a function
```c
Datum
StringToNumeric(char* str) {
	// ...
}
```

And then have that function call the actual function with a PostgresFunctionGuard and name that actual function `StringToNumeric_C`. Similarly to how `GetBaseDuckColumnType` is implemented.
https://github.com/duckdb/pg_duckdb/blob/a8425cb1d5599312d24f4e6f9cd9aad1c1a2bee9/src/pg/types.cpp#L35-L54